Implement role manager default browser dialog in welcome page

aitorvs opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: https://app.asana.com/0/72649045549333/1199338275364983/f
Tech Design URL: 
CC: 

**Description**:
Use the role manager dialog during onboarding welcome page to prompt the user to select us as default browser

* Added the `DefaultRoleBrowserDialogExperiment` that create the intent for the role manager and use it in welcome page

* Skip the "default browser page" regardless of user selection


**Steps to test this PR**:
Note: Use variant `zu`

==Test case 1==
1. Device API >=29
2. Fresh install the app and launch it
3. In Welcome page press "let's do it"
4. Select ddg as default browser in the dialog
5. Verify app proceeds to main browser page
6. Verify pixel `m_db_s` with `fo` param set as true
7. Verify browser is set as default

==Test case 2==
1. Device API >=29
2. Fresh install the app and launch it
3. In Welcome page press "let's do it"
4. Press "Cancel" in the dialog
5. Verify app proceeds to main browser page
6. Verify pixel `m_db_ns` with `fo` param set as true
7. Verify browser is NOT set as default

==Test case 3==
1. Device API <29
2. Fresh install the app and launch it
3. In Welcome page press "let's do it"
4. Verify onboarding proceeds unmodified

==Test case 4==
1. Device with any API
2. Upgrade the app and launch it (onboarding should have been done already)
3. Verify onboarding is skipped and user taken to main browser page

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

marcosholgado commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePage.kt:
```diff
@@ -57,6 +81,23 @@ class WelcomePage : OnboardingPageFragment() {
         typingAnimation?.cancel()
     }
 
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
```
I know this is an experiment but I feel this page could have a view model so this logic could leave there and be tested ala `DefaultBrowserPageViewModel`. 

How about creating a clean up task in Asana to create that view model, test it, etc? If we end up deciding to release to everyone we can do that so, otherwise we delete the task.

aitorvs replied:
Done

marcosholgado reviewed (changes requested):
This is looking good! I've added some comments and I also found a few problems.

If I select our browser as default I get the pixel `m_db_s with params: {fo=true}` as expected but I also get another one `m_db_s with params: {fo=true}`. This double pixel will mess up the current pixel stats and invalidate them when the users set the browser as default out of the onboarding.

I think the above is because you forgot to set the default browser to true `installStore.defaultBrowser = true` so `DefaultBrowserObserver` is kicking in and sending another pixel. Same applies when the result is canceled we should be setting that value to false. Have a look at `fireDefaultBrowserPixelAndResetTimesPressedJustOnce()` for an example. 
We do something different there as well and rather than check the intent result we check if the browser has indeed been set as default I guess just to be sure the browser has actually been set as default.

aitorvs commented on app/src/main/java/com/duckduckgo/app/global/install/AppInstallStore.kt:
```diff
@@ -61,5 +67,6 @@ class AppInstallSharedPreferences @Inject constructor(private val context: Conte
         const val KEY_TIMESTAMP_UTC = "INSTALL_TIMESTAMP_UTC"
         const val KEY_WIDGET_INSTALLED = "KEY_WIDGET_INSTALLED"
         const val KEY_DEFAULT_BROWSER = "KEY_DEFAULT_BROWSER"
+        private const val ROLE_MANAGER_DIALOG_KEY = "ROLE_MANAGER_DIALOG_KEY"
```
Donw

aitorvs commented on app/src/main/java/com/duckduckgo/app/global/DefaultRoleBrowserDialogExperiment.kt:
```diff
@@ -0,0 +1,62 @@
+
+    fun shouldShowExperiment(): Boolean {
+        return variantManager.getVariant().hasFeature(VariantManager.VariantFeature.SetDefaultBrowserDialog) &&
+            appInstallStore.newDefaultBrowserDialogCount < 2
```
done

aitorvs commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePage.kt:
```diff
@@ -40,12 +51,25 @@ class WelcomePage : OnboardingPageFragment() {
+            if (intent != null) {
+                startActivityForResult(intent, DEFAULT_BROWSER_ROLE_MANAGER_DIALOG)
+            } else {
+                onContinuePressed()
```
done

aitorvs commented on app/src/main/java/com/duckduckgo/app/statistics/VariantManager.kt:
```diff
@@ -52,12 +55,21 @@ interface VariantManager {
+            ),
+
+            // Role manager default browser dialog
+            Variant(key = "zt", weight = 1.0, features = emptyList(), filterBy = { isAtLeastApiVersion(Q) }),
```
Added test

aitorvs commented:
@marcosholgado I think I addressed all comments

cmonfortep reviewed (changes requested):
When testing the behavior in <29 api devices, I've found there original behavior has changed.

Develop:
If the user goes to settings and does not select DDG, we don't navigate to the browser.

This PR:
However, when testing this PR, user navigates to the browser. See gif: 

![default_browser](https://user-images.githubusercontent.com/4747865/100618590-54d0c100-331c-11eb-9edb-8b67886035c3.gif)

cmonfortep commented on app/src/main/java/com/duckduckgo/app/onboarding/di/WelcomePageModule.kt:
```diff
@@ -0,0 +1,45 @@
+import dagger.Provides
+
+@Module
+class WelcomePageModule {
```
One question: I see this new approach and I want to know more. would you suggest to create `@Module` per Activity/Fragment? why that instead of per feature? Asking that because we have `OnboardingModule` and I think this could be maybe there?

aitorvs replied:
The setup we currently have has every DI instance in one single component (scope) which is the app scope. However, we are creating one component (scope) per activity/fragment that are completely empty.
This means that everything in DI is visible to every component, even dependencies that are very specific to say a activity/fragment, as they are placed in the app scope. which is kinda anti-pattern for DI.

I thought to take baby steps in this PR to a better DI organization. In this, `WelcomePageModule` has only the dependencies for the welcome Page (eg. `welcomePageViewModelFactory`. Nobody else should see those dependencies in DI or be able to get them. And `WelcomePageModule` is added to the `WelcomePage` scope.

Hope this clarifies

cmonfortep commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingActivity.kt:
```diff
@@ -38,7 +42,9 @@ class OnboardingActivity : DuckDuckGoActivity() {
         val next = viewPager.currentItem + 1
-        if (next < viewPager.adapter!!.count) {
+        if (next < viewPager.adapter!!.count &&
+            !variantManager.getVariant().hasFeature(VariantManager.VariantFeature.SetDefaultBrowserDialog)
```
Correct me if I'm wrong, but I think this condition here assumes that between the First Onboarding screen and the browser Activity there's only 1 screen, and that screen can be skipped when feature `SetDefaultBrowserDialog` is true. 

I think we should avoid having this condition here. What about just not adding pages to the adapter based on that variant feature, and then we can just keep the logic as it was?

aitorvs replied:
Agree. I will move the logic into the `DefaultBrowserDetector`

cmonfortep commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/page/OnboardingPageFragment.kt:
```diff
@@ -32,6 +34,11 @@ abstract class OnboardingPageFragment : Fragment() {
     override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
         inflater.inflate(layoutResource(), container, false)
 
+    override fun onAttach(context: Context) {
```
do we need this here?

aitorvs replied:
I can move it to the `WelcomePage` fragment.

cmonfortep commented on app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePage.kt:
```diff
@@ -24,28 +26,79 @@ import android.view.WindowManager
+    override fun onStart() {
+        super.onStart()
+
+        lifecycleScope.launch {
```
This new pattern looks great, and it remembers me of old good RxJava days :D

In any case, I would recommend that before introducing new patterns we could discuss them internally. Could you do a brief summary of this approach at our next mobile meeting? Thanks!

aitorvs commented:
> When testing the behavior in <29 api devices, I've found there original behavior has changed.
> 
> If the user goes to settings and does not select DDG, we don't navigate to the browser. However, when testing this PR, user navigates to the browser. See gif:
> 
> ![default_browser](https://user-images.githubusercontent.com/4747865/100618590-54d0c100-331c-11eb-9edb-8b67886035c3.gif)

What is the config of the app. Did you force any variant?

cmonfortep commented:
> > When testing the behavior in <29 api devices, I've found there original behavior has changed.
> > If the user goes to settings and does not select DDG, we don't navigate to the browser. However, when testing this PR, user navigates to the browser. See gif:
> 
> What is the config of the app. Did you force any variant?

Yes, sorry. I was using emulator api 28 with `zu` variant. Haven't test the other variant yet, so I can't tell you if it's also happening in there.

cmonfortep commented on app/src/main/java/com/duckduckgo/app/global/DefaultRoleBrowserDialogExperiment.kt:
```diff
@@ -0,0 +1,73 @@
+        return null
+    }
+
+    override fun shouldShowExperiment(): Boolean {
```
question: is there any way to reproduce this in our app?

aitorvs replied:
Not right now, because we are adding the new (role manager) dialog just to onboarding. So the user will only see it once. It will never happen that count is above `1`
I placed this as a safeguard in case we want to use it also in settings in the near future--which was discussed in the near future.

The comment in that method gives the background as to why that safeguard is there.

cmonfortep reviewed (commented):
Looks great, just a few questions from my side.

aitorvs commented:
> > > When testing the behavior in <29 api devices, I've found there original behavior has changed.
> > > If the user goes to settings and does not select DDG, we don't navigate to the browser. However, when testing this PR, user navigates to the browser. See gif:
> > 
> > 
> > What is the config of the app. Did you force any variant?
> 
> Yes, sorry. I was using emulator api 28 ~with `zu` variant~, default variant (I was forcing zu but since I'm api 28, I was assigned to default variant). Haven't tested the other variant yet, so I can't tell you if it's also happening in there.

I found and solved the issue. I've also addressed all comments in https://github.com/duckduckgo/Android/pull/1026/commits/832c8703ab237b3e2425ad3bceb803bc16c09f1a. Let me know if something is amiss.

cmonfortep commented:
When testing the default variant, found a similar bug to the one reported in my previous review.

Bug:
- User selects DDG as default browser via settings
- goes back to the app
- user stays in the default browser onboarding screen

Expected:
- User selects DDG as default browser via settings
- goes back to the app
- we move the user to the Browser activity

![settings_bug](https://user-images.githubusercontent.com/4747865/101210013-149f7480-3675-11eb-9e6e-468d5e2fe8bb.gif)


Tes rest of scenarios work as described 👍

cmonfortep reviewed (commented):
Code looks great to me, thanks for the changes. We should check the bug reported when testing scenario 3.

Apart from that, something I would like to see is some conversations with the rest of the team about the unidirectional pattern we are using in this PR. I think we should share it so others have the chance of asking questions, and be aligned here.
There's also a comment in this PR about changes in our dagger setup. I really appreciate the will of improving our codebase, but I think we are missing some knowledge sharing here. I think everyone will benefit from your expertise in the DI field, so I will really encourage you to share what you have in mind and how can we improve.

I'm out next week, so I'm passing this PR back to @marcosholgado. Thanks and great work!

aitorvs commented:
> When testing the default variant, found a similar bug to the one reported in my previous review.
> 
> Bug:
> 
> * User selects DDG as default browser via settings
> * goes back to the app
> * user stays in the default browser onboarding screen
> 
> Expected:
> 
> * User selects DDG as default browser via settings
> * goes back to the app
> * we move the user to the Browser activity
> 
> Tes rest of scenarios work as described 👍

I tried to repro it and could not. See attached

![screen-20201204-203956_2](https://user-images.githubusercontent.com/6786896/101213008-9e4d4300-3671-11eb-89f2-b0261292b38c.gif)

Do you have any more info that might help reproducing this issue?

marcosholgado reviewed (approved):
Good job! I've added two minor comments related to a couple of tests, feel free to change them or not.

aitorvs commented on app/src/androidTest/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModelTest.kt:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Test
+    fun whenOnPrimaryCtaClickedAndTreatedXpAndNoShowThenFinish() = coroutineRule.runBlocking {
```
Makes total sense, done

aitorvs commented on app/src/androidTest/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModelTest.kt:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Test
+    fun whenOnPrimaryCtaClickedAndTreatedXpNullIntentThenFireAndFinish() = coroutineRule.runBlocking {
```
Makes total sense, done
